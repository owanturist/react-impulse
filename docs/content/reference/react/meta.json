{
  "pages": [
    "use-scope",
    "use-scoped",
    "use-scoped-memo",
    "use-scoped-callback",
    "use-scoped-effect",
    "use-scoped-layout-effect"
  ]
}
