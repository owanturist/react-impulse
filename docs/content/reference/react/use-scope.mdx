---
title: useScope
description: API documentation for the useScope React hook
---

## Overview

The `useScope` hook returns a reactive scope tied to the component lifecycle.

## Function Signature

```typescript
function useScope(): Scope
```

## Returns

A `Scope` instance that triggers component re-renders when read impulses change.

## Description

`useScope` provides a scope that enables reactive impulse reads within React components. When impulses are read within this scope, the component automatically re-renders when those impulses change.

## Examples

```typescript
function Counter() {
  const scope = useScope()
  const impulse = Impulse(0)

  return (
    <div>
      Count: {impulse.getValue(scope)}
    </div>
  )
}
```