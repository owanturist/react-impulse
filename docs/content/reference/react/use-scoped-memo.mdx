---
title: useScopedMemo
description: API documentation for the useScopedMemo React hook
---

## Overview

The `useScopedMemo` hook memoizes computations that depend on impulses, similar to `React.useMemo` but with reactive scope support.

## Function Signature

```typescript
function useScopedMemo<T>(
  factory: (scope: Scope) => T,
  dependencies: DependencyList,
): T
```

## Parameters

- `factory`: A function that receives a `Scope` and computes the memoized value
- `dependencies`: An array of dependencies that control when the memoization is invalidated

## Returns

The memoized value of type `T`.

## Description

`useScopedMemo` combines impulse reactivity with React's memoization. The factory function is re-executed when either the dependencies change or when any impulses read within the scope change.

## Examples

```typescript
const memoizedValue = useScopedMemo(
  (scope) => expensiveComputation(impulse.getValue(scope)),
  [someDependency],
)
```
