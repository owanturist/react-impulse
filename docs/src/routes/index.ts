import { createFileRoute } from "@tanstack/react-router"

import { Home } from "~/pages/home"

export const Route = createFileRoute("/")({
  component: Home,
})
