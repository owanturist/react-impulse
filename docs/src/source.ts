export { create, docs as source } from "../source.generated"
