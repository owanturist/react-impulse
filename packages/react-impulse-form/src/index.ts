export * from "./impulse-form"

export * from "./impulse-form-unit"
export * from "./impulse-form-shape"
export type * from "./impulse-form-meta"
export * from "./impulse-form-list"
export * from "./impulse-form-switch"
export * from "./impulse-form-optional"

export type * from "./result"
export type * from "./zod-like-schema"
export type { ValidateStrategy } from "./validate-strategy"
