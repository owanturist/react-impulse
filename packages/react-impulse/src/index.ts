export type { DependencyList } from "./dependencies"
export type { Compare, Destructor } from "./utils"
export {
  type ImpulseOptions,
  type TransmittingImpulseOptions,
  type ReadonlyImpulse,
  type ImpulseGetter,
  Impulse,
} from "./Impulse"
export type { Scope } from "./Scope"
export { untrack } from "./untrack"
export { batch } from "./batch"
export { batch as tap } from "./batch"
export { subscribe } from "./subscribe"
export { useScope } from "./useScope"
export { type UseScopedOptions, useScoped } from "./useScoped"
export { useScopedMemo } from "./useScopedMemo"
export { useScopedCallback } from "./useScopedCallback"
export { useScopedEffect } from "./useScopedEffect"
export { useScopedLayoutEffect } from "./useScopedLayoutEffect"
