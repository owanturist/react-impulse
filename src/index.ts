export type { Compare } from "./utils"
export type { Scope } from "./Scope"
export { Impulse } from "./Impulse"
export { tap } from "./tap"
export { batch } from "./batch"
export { subscribe } from "./subscribe"
export { watch } from "./watch"
export { useImpulse } from "./useImpulse"
export { useImpulseValue } from "./useImpulseValue"
export { useImpulseMemo } from "./useImpulseMemo"
export { useImpulseEffect, useImpulseLayoutEffect } from "./useImpulseEffects"
export { useWatchImpulse } from "./useWatchImpulse"
