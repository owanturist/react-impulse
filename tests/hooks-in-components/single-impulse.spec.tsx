import React from "react"
import { act, render, screen, fireEvent } from "@testing-library/react"

import { Impulse, useImpulseValue } from "../../src"
import { Counter } from "../common"

import { withinNth } from "./common"

describe("single impulse", () => {
  const GetterComponent: React.FC<{
    counter: Impulse<Counter>
    onRender: VoidFunction
  }> = ({ counter, onRender }) => {
    const { count } = useImpulseValue(counter)

    return (
      <React.Profiler id="test" onRender={onRender}>
        <span data-testid="getter">{count}</span>
      </React.Profiler>
    )
  }

  const SetterComponent: React.FC<{
    counter: Impulse<Counter>
    onRender: VoidFunction
  }> = ({ counter, onRender }) => (
    <React.Profiler id="test" onRender={onRender}>
      <div data-testid="setter">
        <button
          type="button"
          data-testid="increment"
          onClick={() => counter.setValue(Counter.inc)}
        />
        <button
          type="button"
          data-testid="reset"
          onClick={() => counter.setValue({ count: 0 }, Counter.compare)}
        />
      </div>
    </React.Profiler>
  )

  const SingleSetterSingleGetter: React.FC<{
    counter: Impulse<Counter>
    onRootRender: VoidFunction
    onGetterRender: VoidFunction
    onSetterRender: VoidFunction
  }> = ({ counter, onRootRender, onGetterRender, onSetterRender }) => (
    <>
      <React.Profiler id="test" onRender={onRootRender} />

      <GetterComponent counter={counter} onRender={onGetterRender} />
      <SetterComponent counter={counter} onRender={onSetterRender} />
    </>
  )

  it("Single Setter / Getter", () => {
    const counter = Impulse.of({ count: 0 })
    const onRootRender = vi.fn()
    const onGetterRender = vi.fn()
    const onSetterRender = vi.fn()

    render(
      <SingleSetterSingleGetter
        counter={counter}
        onRootRender={onRootRender}
        onGetterRender={onGetterRender}
        onSetterRender={onSetterRender}
      />,
    )

    // check initial value
    expect(onRootRender).toHaveBeenCalledOnce()
    expect(onSetterRender).toHaveBeenCalledOnce()
    expect(onGetterRender).toHaveBeenCalledOnce()
    expect(screen.getByTestId("getter")).toHaveTextContent("0")
    vi.clearAllMocks()

    // increment by from the component
    fireEvent.click(screen.getByTestId("increment"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).toHaveBeenCalledOnce()
    expect(screen.getByTestId("getter")).toHaveTextContent("1")
    vi.clearAllMocks()

    // increment from the outside
    act(() => counter.setValue(Counter.inc))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).toHaveBeenCalledOnce()
    expect(screen.getByTestId("getter")).toHaveTextContent("2")
    vi.clearAllMocks()

    // reset from the component
    fireEvent.click(screen.getByTestId("reset"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).toHaveBeenCalledOnce()
    expect(screen.getByTestId("getter")).toHaveTextContent("0")
    vi.clearAllMocks()

    // reset second time in a row
    fireEvent.click(screen.getByTestId("reset"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).not.toHaveBeenCalled()
    expect(screen.getByTestId("getter")).toHaveTextContent("0")
    vi.clearAllMocks()

    // increment twice in a row
    fireEvent.click(screen.getByTestId("increment"))
    fireEvent.click(screen.getByTestId("increment"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).toHaveBeenCalledTimes(2)
    expect(screen.getByTestId("getter")).toHaveTextContent("2")
    vi.clearAllMocks()

    // increment twice in a row from the outside
    act(() => {
      counter.setValue(Counter.inc)
      counter.setValue(Counter.inc)
    })
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onSetterRender).not.toHaveBeenCalled()
    expect(onGetterRender).toHaveBeenCalledOnce()
    expect(screen.getByTestId("getter")).toHaveTextContent("4")
  })

  const MultipleSetterMultipleGetter: React.FC<{
    counter: Impulse<Counter>
    onRootRender: VoidFunction
    onFirstGetterRender: VoidFunction
    onSecondGetterRender: VoidFunction
    onFirstSetterRender: VoidFunction
    onSecondSetterRender: VoidFunction
  }> = ({
    counter,
    onRootRender,
    onFirstGetterRender,
    onSecondGetterRender,
    onFirstSetterRender,
    onSecondSetterRender,
  }) => (
    <>
      <React.Profiler id="test" onRender={onRootRender} />
      <GetterComponent counter={counter} onRender={onFirstGetterRender} />
      <GetterComponent counter={counter} onRender={onSecondGetterRender} />
      <SetterComponent counter={counter} onRender={onFirstSetterRender} />
      <SetterComponent counter={counter} onRender={onSecondSetterRender} />
    </>
  )

  it("Multiple Setters / Getters", () => {
    const counter = Impulse.of({ count: 0 })
    const onRootRender = vi.fn()
    const onFirstGetterRender = vi.fn()
    const onSecondGetterRender = vi.fn()
    const onFirstSetterRender = vi.fn()
    const onSecondSetterRender = vi.fn()

    render(
      <MultipleSetterMultipleGetter
        counter={counter}
        onRootRender={onRootRender}
        onFirstGetterRender={onFirstGetterRender}
        onSecondGetterRender={onSecondGetterRender}
        onFirstSetterRender={onFirstSetterRender}
        onSecondSetterRender={onSecondSetterRender}
      />,
    )

    // check initial value
    expect(onRootRender).toHaveBeenCalledOnce()
    expect(onFirstSetterRender).toHaveBeenCalledOnce()
    expect(onSecondSetterRender).toHaveBeenCalledOnce()
    expect(onFirstGetterRender).toHaveBeenCalledOnce()
    expect(onSecondGetterRender).toHaveBeenCalledOnce()
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()
    vi.clearAllMocks()

    // increment from the first component
    fireEvent.click(withinNth("setter", 0).getByTestId("increment"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onFirstSetterRender).not.toHaveBeenCalled()
    expect(onSecondSetterRender).not.toHaveBeenCalled()
    expect(onFirstGetterRender).toHaveBeenCalledOnce()
    expect(onSecondGetterRender).toHaveBeenCalledOnce()
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()
    vi.clearAllMocks()

    // increment from the second component
    fireEvent.click(withinNth("setter", 1).getByTestId("increment"))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onFirstSetterRender).not.toHaveBeenCalled()
    expect(onSecondSetterRender).not.toHaveBeenCalled()
    expect(onFirstGetterRender).toHaveBeenCalledOnce()
    expect(onSecondGetterRender).toHaveBeenCalledOnce()
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()
    vi.clearAllMocks()

    // increment from the outside
    act(() => counter.setValue(Counter.inc))
    expect(onRootRender).not.toHaveBeenCalled()
    expect(onFirstSetterRender).not.toHaveBeenCalled()
    expect(onSecondSetterRender).not.toHaveBeenCalled()
    expect(onFirstGetterRender).toHaveBeenCalledOnce()
    expect(onSecondGetterRender).toHaveBeenCalledOnce()
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()
  })
})
