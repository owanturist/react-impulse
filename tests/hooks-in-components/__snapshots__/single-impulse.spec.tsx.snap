// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`single impulse > multiple Setters / Getters 1`] = `
[
  <span
    data-testid="getter"
  >
    0
  </span>,
  <span
    data-testid="getter"
  >
    0
  </span>,
]
`;

exports[`single impulse > multiple Setters / Getters 2`] = `
[
  <span
    data-testid="getter"
  >
    1
  </span>,
  <span
    data-testid="getter"
  >
    1
  </span>,
]
`;

exports[`single impulse > multiple Setters / Getters 3`] = `
[
  <span
    data-testid="getter"
  >
    2
  </span>,
  <span
    data-testid="getter"
  >
    2
  </span>,
]
`;

exports[`single impulse > multiple Setters / Getters 4`] = `
[
  <span
    data-testid="getter"
  >
    3
  </span>,
  <span
    data-testid="getter"
  >
    3
  </span>,
]
`;
